service/src/java/org/apache/hive/service/server/HiveServer2.java

```diff
@@ -1204,17 +1205,21 @@ private static void startHiveServer2() throws Throwable {
   private void maybeStartCompactorThreads(HiveConf hiveConf) throws Exception {
     if (MetastoreConf.getVar(hiveConf, MetastoreConf.ConfVars.HIVE_METASTORE_RUNWORKER_IN).equals("hs2")) {
       int numWorkers = MetastoreConf.getIntVar(hiveConf, MetastoreConf.ConfVars.COMPACTOR_WORKER_THREADS);
-      List<Map.Entry<String, String>> entries = hiveConf.getMatchingEntries(Constants.COMPACTION_POOLS_PATTERN);
+      Map<String, Integer> customPools = CompactorUtil.getPoolConf(hiveConf);
 
       StringBuilder sb = new StringBuilder(2048);
       sb.append("This HS2 instance will act as Compactor with the following worker pool configuration:\n");
       sb.append("Global pool size: ").append(numWorkers).append("\n");
 
       LOG.info("Initializing the compaction pools with using the global worker limit: {} ", numWorkers);
-      while (numWorkers > 0 && entries.size() > 0) {
-        Map.Entry<String, String> entry = entries.remove(0);
-        String poolName = entry.getValue();
-        int poolWorkers = hiveConf.getInt(entry.getKey(), 0);
+      for (Map.Entry<String, Integer> customPool: customPools.entrySet()) {
+        
+        if (numWorkers <= 0) {
```

Comment:
why do we need this check inside the loop?

Reply:
Because in each iteration it subtracts `poolWorkers` from `numWorkers` and the loop execution needs to stop when there are no more workers left. `poolWorkers` is the number of custom pool workers and `numWorkers` is the global limit of workers.

```
String poolName = customPool.getKey();
int poolWorkers = customPool.getValue();
...
numWorkers -= poolWorkers;
```